### ISSUE TITLE
TBBConfig.cmake should not add '-isystem /usr/local' to gcc's CXX_INCLUDES flag

### ISSUE THREAD

**Issue by ldorau** [oneapi-src/oneTBB#195]
TBBConfig.cmake:
https://github.com/intel/tbb/blob/tbb_2019/cmake/templates/TBBConfig.cmake.in#L57
sets the property `INTERFACE_INCLUDE_DIRECTORIES` to `${CMAKE_CURRENT_LIST_DIR}/../../../include` (which equals `/usr/include`).
It causes that
` -isystem /usr/lib64/cmake/TBB/../../../include `
(equal `-isystem /usr/include`)
is being added to gcc's `CXX_INCLUDES` flag, what is wrong and causes a compilation error - see https://github.com/pmem/libpmemobj-cpp/issues/508

**Comment by ldorau** [comment]
Seen the fix: https://github.com/pmem/libpmemobj-cpp/pull/514

**Comment by ldorau** [comment]
It is a workaround, not a fix!

**Comment by Aldrog** [comment]
From discussion of opencv/opencv#15795 I understood it that cmake will filter out directories like this if they'll be normalized. I think you can get a normalized path if you construct it with `file` cmake command.

**Comment by Aldrog** [comment]
This one works:
```cmake
get_filename_component(COMPONENT_INCLUDE_DIRECTORY "${CMAKE_CURRENT_LIST_DIR}/@TBB_INC_REL_PATH@" ABSOLUTE)
set_target_properties(TBB::${_tbb_component} PROPERTIES
                    INTERFACE_INCLUDE_DIRECTORIES "${COMPONENT_INCLUDE_DIRECTORY}")
```

**Comment by tbbdev** [comment]
@Aldrog thank you for the proposal!
It was applied in https://github.com/intel/tbb/blob/v2020.0/cmake/templates/TBBConfig.cmake.in#L50

**Comment by RuRo** [comment]
I am having some issues with building `opencv-cuda` with `tbb v2020.1` on Arch Linux. I am not too sure, how `${CMAKE_CURRENT_LIST_DIR}/@TBB_INC_REL_PATH@` works, but it seems, that in my case `INTERFACE_INCLUDE_DIRECTORIES` got resolved to just `/include`.

The error is
```markdown
CMake Error in modules/core/CMakeLists.txt:
  Imported target "tbb" includes non-existent path

    "/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.
```

And the related template lines in `/usr/lib/cmake/TBB/TBBConfig.cmake` got resolved to
```cmake
            get_filename_component(_tbb_include_dir "${CMAKE_CURRENT_LIST_DIR}/../../../include" ABSOLUTE)
            set_target_properties(TBB::${_tbb_component} PROPERTIES
                                  INTERFACE_INCLUDE_DIRECTORIES "${_tbb_include_dir}")
            unset(_tbb_include_dir)
```

I am not 100% sure, that this is `tbb`s fault, but this issue seemed related.

**Comment by alalek** [comment]
Try to debug CMake scripts: `cmake ... --trace-expand`

**Comment by RuRo** [comment]
The full `--trace-expand --trace-redirect=trace.log` file is 1.5GB (157M gzip compressed), so I am not going to post it here, unless you really need the whole thing. The relevant lines seem to be
```
/lib64/cmake/TBB/TBBConfig.cmake(58):  add_library(TBB::tbb SHARED IMPORTED )
/lib64/cmake/TBB/TBBConfig.cmake(60):  get_filename_component(_tbb_include_dir /lib64/cmake/TBB/../../../include ABSOLUTE )
/lib64/cmake/TBB/TBBConfig.cmake(61):  set_target_properties(TBB::tbb PROPERTIES INTERFACE_INCLUDE_DIRECTORIES /include )
/lib64/cmake/TBB/TBBConfig.cmake(63):  unset(_tbb_include_dir )
```
so apparently, `${CMAKE_CURRENT_LIST_DIR}` is `/lib64/cmake/TBB` during the invocation? Not sure, why is that the case. If you need the whole 1.5GB file or want me to grep for something in it, feel free to ask.

**Comment by alalek** [comment]
This part of CMake's log is enough.

Please dump this from your system:
```
ls -al /usr
ls -al /lib64
```

How is TBB installed/build? AUR?

**Comment by RuRo** [comment]
<details>

```
$ /bin/ls -al /usr
total 608
drwxr-xr-x  12 root root   4096 Feb  3 19:18 .
drwxr-xr-x  19 root root   4096 Dec 25 13:19 ..
drwxr-xr-x   7 root root 155648 Feb  3 13:44 bin
drwxr-xr-x   3 root root   4096 Jan  7 16:20 doc
drwxr-xr-x   3 root root   4096 Oct  5  2018 etc
drwxr-xr-x 616 root root  69632 Feb  3 13:44 include
drwxr-xr-x 269 root root 294912 Feb  3 13:44 lib
drwxr-xr-x  42 root root  45056 Feb  1 20:21 lib32
lrwxrwxrwx   1 root root      3 Nov 20 01:24 lib64 -> lib
drwxr-xr-x  13 root root   4096 Dec 21  2018 local
drwxr-xr-x   3 root root   4096 Nov 15 19:10 man
lrwxrwxrwx   1 root root      3 Nov 20 01:24 sbin -> bin
drwxr-xr-x 406 root root  12288 Feb  3 13:44 share
drwxr-xr-x   2 root root   4096 Jun 21  2017 src

$ /bin/ls -al /lib64
lrwxrwxrwx 1 root root 7 Nov 20 01:24 /lib64 -> usr/lib
```

</details>

TBB was installed from [extra/intel-tbb](https://www.archlinux.org/packages/extra/x86_64/intel-tbb/).

**Comment by alalek** [comment]
`/lib64` is a symlink on normal /usr/lib. No idea why it is handled first (details are somewhere in gcc or cmake internals).

@RuRo Could you try to change `TBBConfig.cmake` file on your system in this way?

```.diff
+get_filename_component(_tbb_CMAKE_CURRENT_LIST_DIR "${CMAKE_CURRENT_LIST_DIR}" REALPATH)  # resolve symlinks
+message("DEBUG: TBB CMAKE_CURRENT_LIST_DIR: ${_tbb_CMAKE_CURRENT_LIST_DIR}")
 get_filename_component(_tbb_include_dir "${_tbb_CMAKE_CURRENT_LIST_DIR}/../../../include" ABSOLUTE)
+message("DEBUG: TBB inlcude: ${_tbb_include_dir}")
 set_target_properties(TBB::${_tbb_component} PROPERTIES
                                  INTERFACE_INCLUDE_DIRECTORIES "${_tbb_include_dir}")
 unset(_tbb_include_dir)
```

**Comment by RuRo** [comment]
Okay, resolving symlinks in `TBBConfig.cmake` like you suggested worked for me.

**Comment by alalek** [comment]
@RuRo Thanks for check!

Perhaps this patch should work too (ABSOLUTE => REALPATH):
```.diff
-get_filename_component(_tbb_include_dir "${CMAKE_CURRENT_LIST_DIR}/../../../include" ABSOLUTE)
+get_filename_component(_tbb_include_dir "${CMAKE_CURRENT_LIST_DIR}/../../../include" REALPATH)
```

**Comment by RuRo** [comment]
Just changing `ABSOLUTE` to `REALPATH` doesn't seem to work.

**Comment by AlexVeprev** [comment]
@alalek, @RuRo, thank you for so detailed investigation! I checked it on my side too. The fix is planned to be implemented in one of future TBB releases.

**Comment by tbbdev** [comment]
The fix was implemented in v2020.2: https://github.com/oneapi-src/oneTBB/blob/v2020.2/cmake/templates/TBBConfig.cmake.in#L60
